104.md

```diff
@@ -0,0 +1,205 @@
+
+```json
+{
+    "kind": 444,
+    "created_at": <unix timestamp in seconds>,
+    "pubkey": <pubkey of the sender>,
+    "content": <ciphertext of the message symmetrically encrypted with chain/message key>,
+    "tags": [
+        ["p", <pubkey of the recipient>],
+        ["dh_sending", <current DH sending pubkey>],
+        ["current_index", <number of current message in the current message chain>],
+        ["previous_length", <length of previous message chain>]
+    ],
+    "sig": "" // No signature because these are only sent inside gift wraps.
+}
+```
+
+#### Required Tags
+
+- `"p"`: References the recipient
+- `"dh_sending"`: The ephemeral pubkey used in the current Diffie-Hellman ratchet.
+- `"current_index"`: The index of the current message in the current message chain. For handling out of order messages.
+- `"previous_length"`: The length of the previous message chain. For handling out of order messages.
+
+Clients can use any other standard tags they like, for example, `"a"` or `"e"` tags to reference previous messages.
+
+### Prekey Event
+
+The Prekey event is a simple replaceable event where a user publishes their signed prekey that is required to initiate the Diffie-Hellman key exchange to create a conversation request event. Including a signature on this event proves that the holder of the user's main nostr key (which signs the `kind: 10443` event) also has control of the private key associated with the prekey.
+
+In most cases, clients that implement this NIP will manage the creation and rotation of this key on a regular basis. It's recommended that clients do so interactively with user consent in order to avoid overwriting prekeys created by other clients. It is also recommended that clients rotate this key on a regular basis.
+
+```json
+{
+    "kind": 10443,
+    "created_at": <unix timestamp in seconds>,
+    "pubkey": <pubkey>,
+    "content": <hex encoded pubkey of the prekey>,
+    "tags": [
+        ["prekey_sig", <signature generated from hex encoded pubkey of the prekey>]
+    ],
+    "sig": <signed normally with the user's main nostr pubkey>
+}
+```
+
+The `"prekey_sig"` tag value is a Schnorr signature of the sha256 hashed value of the prekey's pubkey, signed with the prekey's private key.
+
+```js
+const privKey = schnorr.utils.randomPrivateKey();
+const pubKey = bytesToHex(schnorr.getPublicKey(privKey));
+
+const prekeySig = bytesToHex(
+  schnorr.sign(bytesToHex(sha256(utf8Encoder.encode(pubKey)), privKey))
+);
+
+const prekeyTag = ["prekey_sig", prekeySig];
+```
+
+## Known trade-offs & open questions
+
+1. Messages are decrypted and stored on each device/client pairing. This means that there is no way to load and decrypt historic messages directly from relays on a different client or device. However, you can create backups of your decrypted conversations that could be loaded into another device (this is beyond the scope of this NIP).
```

Comment:
No, in most cases that I've seen there is no historic conversation syncing via the protocol. With some of them there is a way to download/backup chat history and then load that into a new client (Signal does this) but it's not syncing conversations between devices ever.

One thought I had about this was that if Blossom becomes something then it would be a way that we could sync this chat data but I think that's still too early to tell...

Reply:
https://conversations.im/ suggests "If you are installing Conversations on a new device or catching up after being offline for a while, Conversations will use [XEP-0313: Message Archive Management](https://xmpp.org/extensions/xep-0313.html) to fetch the message history from your server.". Do they mean that it is a new device with the same private key loaded from an old device?

Reply:
Signal does not address the issue. Signal can't download historical messages and doesn't have sync (besides a very trivial one, phone-to-desktop-web that relies on phone-to-web pairing and LAN webservers)

Reply:
> Signal does not address the issue. Signal can't download historical messages and doesn't have sync (besides a very trivial one, phone-to-desktop-web that relies on phone-to-web pairing and LAN webservers)

That's partially true. They don't handle syncing but they do have a mechanism to back up the history and then load that file into a new device (done via bluetooth) https://support.signal.org/hc/en-us/articles/360007059752-Backup-and-Restore-Messages

@AndySchroder No idea. I think we have to approach the syncing question from first principles based on how Nostr works instead of just trying to copy whatever XMPP is doing.